### ISSUE TITLE
Future Work

### ISSUE THREAD

**Issue by daniel-schuermann** [daniel-schuermann/mesa#98]
The purpose of this issue is purely informative and should serve clarification about recurring questions whether ACO will support certain features in future.

- **Supported hardware:**
 - [x] GFX6 (Radeon 7000 series, Radeon 200 series)
 - [x] GFX7 (Radeon 290, Radeon 300 series)
 - [x] GFX8 (Tonga, Fiji, Polaris)
 - [x] GFX9 (Vega)
 - [x] GFX10 (Navi)
 - [x] APUs (Bristol Ridge, Raven Ridge, ...)
 - [ ] ~Intel & Nvidia~ (not possible)

- **Supported shader stages:**
- [x] Vertex shaders
- [x] Tesselation control shaders
- [x] Tesselation evaluation shaders
- [x] Geometry shaders
- [x] Fragment shaders
- [x] Compute shaders

- **Supported APIs:**
- [x] Vulkan
- [ ] OpenGL
- [ ] ~OpenCL~ (not planned for near to mid future)

**Comment by SpookySkeletons** [comment]
Would Intel be open to providing pull requests to the compiler for their GPU's arch?
Or is it some other blocker?

What license is ACO released under? MIT, same as Mesa?

**Comment by daniel-schuermann** [comment]
> Would Intel be open to providing pull requests to the compiler for their GPU's arch?
Or is it some other blocker?

The backend is GCN-only, and trying to port it to some other vendor's ISA basically means to rewrite it from scratch, so no.

> What license is ACO released under? MIT, same as Mesa?

Yes.

**Comment by SR-dude** [comment]
FYI; I just tried a release candidate for 5.3 and it wasn't a pleasant experience.
I only tried _The _Division__  and  the garbage displayed on the screen should
be worrisome to ACO developers.  I had no choice but to downgrade.

**Comment by pendingchaos** [comment]
@SR-dude: Does this issue only happen with ACO?

There was a regression with Vega (fixed in 994dc8eefdce3b7071850087574bb8ba84f6752f) which affected anything using Vulkan (but that shouldn't be affected by the kernel version though).

If it wasn't fixed by that and only happens with ACO, this should probably have it's own issue created.

**Comment by SR-dude** [comment]
I only tested ACO. I was looking into the strange GPU resets with the aforementioned game.

**Comment by daniel-schuermann** [comment]
As ACO is only a compiler backend, it barely interacts with the userspace driver and nothing else. It just translates code from one language to another.

**Comment by ntropy83** [comment]
The point "Supported shader stages", what does it mean exactly? As far as I know current ACO is not leaving out the rendering of tesselation or geometry shaders in a game. 

One last question: why is Intel support not possible? Could that change if they release their dGPUs?

**Comment by pendingchaos** [comment]
> The point "Supported shader stages", what does it mean exactly? As far as I know current ACO is not leaving out the rendering of tesselation or geometry shaders in a game.

It means RADV falls back to LLVM for these shaders because we haven't yet implement the features needed to compile them.

> One last question: why is Intel support not possible? Could that change if they release their dGPUs?

ACO is a AMD specific because it turns a shader in a representation used by many Mesa drivers and common code (NIR) into GCN (and in the future, RDNA) code. So it's extremely AMD specific and supporting Intel CPUs would mean adding backends to ACO (which is already a backend) and turning it's IR into something resembling NIR.

It's role is similar to a LLVM backend and reason why it can't be used for Intel is the same reason why LLVM's X86 backend doesn't support ARM CPUs.

Intel's dGPUs almost certainly won't be using GCN/RDNA, so it won't change when they release them.

**Comment by ntropy83** [comment]
That makes sense, thank you for explaining. 
I am curious to see the end result, keep up the good work. If you ever need some helping hands let me know.

**Comment by Maltahl** [comment]
Is it currently possible to build a version of mesa-git-aco that supports navi10 (GFX10) ?
I dont mind if i have to compile it myself and test it out.

Binaries provided on https://people.freedesktop.org/~agd5f/radeon_ucode/navi10/ are currently broken since 8/8 .
However, the binaries on https://koji.fedoraproject.org/koji/buildinfo?buildID=1314657 works great with mesa-git using llvm 10 but i really wanna test the ACO patched mesa driver on Navi10

**Comment by pendingchaos** [comment]
> Is it currently possible to build a version of mesa-git-aco that supports navi10 (GFX10) ?
> I dont mind if i have to compile it myself and test it out.

Not sure which exact package your talking about, but a sufficiently up-to-date one should be compiled with Navi support.

> Binaries provided on https://people.freedesktop.org/~agd5f/radeon_ucode/navi10/ are currently broken since 8/8 .
> However, the binaries on https://koji.fedoraproject.org/koji/buildinfo?buildID=1314657 works great with mesa-git using llvm 10 but i really wanna test the ACO patched mesa driver on Navi10

It shouldn't matter what firmware you use with ACO, as long as the firmware works.

Note that ACO doesn't yet support Navi, so it will fall back to LLVM.

**Comment by Maltahl** [comment]
> Not sure which exact package your talking about, but a sufficiently up-to-date one should be compiled with Navi support.
> It shouldn't matter what firmware you use with ACO, as long as the firmware works.
> 
> Note that ACO doesn't yet support Navi, so it will fall back to LLVM.

Thanks for the very quick reply.
Good to know but there is still a warning where fallback is not happening on the package from valveaur mesa-aco-git  1:19.2.0_devel.115181.34dd1ddde6f-1

The driver that works for me currently is mesa-git compiled with your llvm (PKGBUILD from tk-glitch on github) so thanks a lot for the llvm build you have @pendingchaos since it is the llvm version that works for me.

**Comment by SR-dude** [comment]
Put your mind at ease. Kernel version 5.3-rc4 is ok and I'm not having any problem with ACO by using that kernel version.

**Comment by shmerl** [comment]
> Put your mind at ease. Kernel version 5.3-rc4 is ok and I'm not having any problem with ACO by using that kernel version.

But aco is still not supporting Navi?

**Comment by vosian** [comment]
The original post says that ACO doesn't yet support Navi, so you'll have to wait a while.

**Comment by daniel-schuermann** [comment]
ACO should now be able to support GFX7 GPUs (Radeon R7 260/360(X) and R9 290/390(X)).
(For upstream support, please check on https://gitlab.freedesktop.org/mesa/mesa/merge_requests/2751 if it was merged.)

**Comment by MEXAHOTABOP** [comment]
gfx7 support can be upstreamed in coming 19.3.0 or 20.0/19.3.1+ only?

**Comment by BNieuwenhuizen** [comment]
I'd expect the gfx7 support to first end up in 20.0.

**Comment by MEXAHOTABOP** [comment]
this mean release somewhere in march 
thanks

**Comment by daniel-schuermann** [comment]
It's a bit unfortunate, but there isn't much we can do about. 19.3 has already been branched off, and only bug fixes are allowed to be backported. New features like this one automatically end up in the next release cycle.

**Comment by tiburcillo** [comment]
allright, does that mean I can use Ernst's or Kisak's PPA (https://launchpad.net/~kisak/+archive/ubuntu/kisak-mesa) or (https://launchpad.net/~ernstp/+archive/ubuntu/mesaaco?field.series_filter=eoan) to test this functionality?

**Comment by daniel-schuermann** [comment]
Both PPAs switched to 19.3, which means no. You can test it usind either https://launchpad.net/~valve-experimental (currently has only Bionic packages..) or wait a bit for it to land and use padoka PPA with RADV_PERFTEST=aco

**Comment by tiburcillo** [comment]
Thanks so much for this work, I'll keep my eyes open for a Mesa 20 PPA or maybe try to compile myself. 
Will report back in Steam Community once I grilled my good old R9 390X.

> Both PPAs switched to 19.3, which means no. You can test it usind either https://launchpad.net/~valve-experimental (currently has only Bionic packages..) or wait a bit for it to land and use padoka PPA with RADV_PERFTEST=aco

**Comment by SpookySkeletons** [comment]
Is it a good idea to add a config option to the meson build to disable llvm dependency requirements completely once ACO is feature complete if amdgpu is the only lib enabled?
Would be very useful for Gentoo down the line.

**Comment by daniel-schuermann** [comment]
I think so, but this will probably not be the case before 20.1 (at least if you also want to build radeonSI).

**Comment by MEXAHOTABOP** [comment]
used aco on gfx7 for a week seems like no problems so far
good job

**Comment by eiglow** [comment]
is GFX6 support something that's coming? my HD 7970 and r5 240 are hanging out for it

**Comment by shmerl** [comment]
Is there any rough ETA for tessellation and geometry shaders support?

**Comment by pendingchaos** [comment]
GFX6, tessellation and geometry shaders are on the TODO list, there isn't really an ETA though

There is a WIP MR for geometry shaders on the freedesktop Gitlab  which seems to work on Vega/Navi, so hopefully geometry shaders isn't too far away (though I expect the holidays to slow down review).

**Comment by SpookySkeletons** [comment]
Thanks for the heads up on the merge.
Will gladly test the gs patches on my GFX8 when it finishes compiling with LTO!

Update: The gs patches seem to work alright on my GFX8 in Mordhau through steamplay! Nothing exploded or crashed

**Comment by marekolsak** [comment]
The section about supported shader stages is ambiguous. You know the vertex shader can be compiled as VS, ES, or LS, and the tess eval shader can be compiled as VS or ES. Also on gfx10, vertex shaders are usually compiled as NGG GS.

**Comment by pendingchaos** [comment]
Yeah, the item on vertex shaders is a little unclear. Vertex shaders as VS is supported but ES/LS (used when geometry or tesselation shaders are used) is unsupported. All kinds of geometry and tesselation shaders are unsupported. NGG is currently unsupported and it falls back to the legacy pipeline

The GS MR implements vertex shaders as ES, merged vertex/geometry shaders and GS copy shaders

**Comment by shmerl** [comment]
And is NGG work planned after tessellation shaders?

**Comment by pendingchaos** [comment]
There isn't really a planned order

If implementing NGG would improve vertex shader performance though, I would prioritize it over tessellation shaders. I haven't gotten around to actually checking this though

**Comment by marekolsak** [comment]
If NGG culling was implemented, it would double primitive throughout, because NGG GS can process primitives 2x faster than the rasterizer. Without culling, you are limited by the rasterizer. The legacy VS can only process primitives at the same rate as the rasterizer.

**Comment by marekolsak** [comment]
The downside of NGG culling is that it consumes more shader resources. The shaders are longer, have to contain culling code and thread compaction code, use LDS, and contain quite a few barriers.

**Comment by SR-dude** [comment]
Is this repository now deprecated? Any updates?

**Comment by shmerl** [comment]
As far as I understood, it's recommended to use Mesa master for any incoming aco updates.

**Comment by daniel-schuermann** [comment]
> Is this repository now deprecated? Any updates?

Let's say it's deprecated, but will still receive updates for a while (probably until mesa 20.0.x). Currently, we use it as testing bed for merge requests, and try to rebase every week or so.

**Comment by baryluk** [comment]
@daniel-schuermann I think it is a good idea to close the issue tracker and not accept new issue reports.

**Comment by daniel-schuermann** [comment]
Upstream has no issue templates yet, so no. At least not yet :)

**Comment by pendingchaos** [comment]
Geometry shader support (from https://gitlab.freedesktop.org/mesa/mesa/merge_requests/2421 ) has been pushed to this repository. It hasn't been upstreamed yet.

Geometry shaders are enabled for all hardware ACO supports. Some of the commit messages mention "GFX8", but geometry shaders work the same on GFX7 and so should be enabled there too (the commits were initially made before GFX7 support was upstreamed IIRC).

**Comment by shmerl** [comment]
Do geometry shaders need any flags to enable wave32 for them like in other cases?

**Comment by pendingchaos** [comment]
When NGG is disabled for geometry shaders (which is currently always the case with both ACO and LLVM, but LLVM will soon use NGG), RADV always uses wave64 for geometry shaders

**Comment by aufkrawall** [comment]
Are LLVM's compile times for GS comparably bad as for others?
Seems like the immediate advantages of having it in ACO are not that distinct, but perhaps it will be beneficial when seen in a bigger picture?
At least I don't notice any breakage either.

**Comment by pendingchaos** [comment]
GS isn't really used that often, so the effect isn't as large as using ACO for VS, FS or CS shaders.

**Comment by baryluk** [comment]
Once the GS (and the dependent MRs) is upstreamed, would it be possible to compile Mesa or at least RADV without LLVM maybe?  (loosing possibly decompilation support, which is mostly useful for debugging right? Or it is also used for other things like shader compiler feedback?).

**Comment by pendingchaos** [comment]
Besides GS, there is still tesselation shaders and GFX6 which currently fall back to LLVM. There are also some features which are safe to disable when ACO is used but are needed for feature parity: 8/16-bit types and NGG

After all that, we use LLVM's disassembler for GFX8+. So an alternative would need to be found or disassembly would have to be disabled (not a huge loss though, it's mostly useful for debugging and RenderDoc)

If it were possible to compile RADV without LLVM, llvmpipe and radeonsi would still need LLVM

**Comment by baryluk** [comment]
@pendingchaos Yeah, I know tesselation shaders still need LLVM.

Not having disassembler is fine for normal users.

But yeah, it is not practical to just RADV, but not radeonsi for some other stuff.

Well, once radeonsi uses aco, it is probably worth looking into it again :)

Thanks.

**Comment by pendingchaos** [comment]
Geometry shader support (without NGG) has been upstreamed.

GFX6 support has also been upstreamed and should be in this repository when it's rebased

**Comment by daniel-schuermann** [comment]
So, sorry for the delay. Now, I rebased and GFX6 and GS should also work for those using the valve ppa.

**Comment by Venemo** [comment]
To those interested: I've started the work on tessellation support for ACO. There is a branch here: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/3964 which should work accross all devices, although so far I've tested with Polaris and Navi. Feedback is welcome, in case anyone wants to check it out.

**Comment by shmerl** [comment]
I'll give it a test later today. Can it be merged on its own for testing, or some other MRs are needed as well?

**Comment by Venemo** [comment]
@shmerl It doesn't need to be merged with anything, you can just checkout the `aco-tess` branch on my gitlab.

**Comment by Venemo** [comment]
@shmerl Fair warning: I checked, and it doesn't improve the frame rates of The Witcher 3 on Navi. Just saying.

**Comment by shmerl** [comment]
I'll check, at least to confirm that it doesn't hang :) Also, I suspect framerate impact can be somewhat resolution specific, at least I already encountered such comparisons in the past.

**Comment by gort818** [comment]
I will be giving it a test on Vega hardware.

**Comment by shmerl** [comment]
@Venemo: posted a result in benchmarks thread (#36). As you anticipated, no performance improvement for TW3, but all worked well in general. llvm11 still has a performance regression.

**Comment by Venemo** [comment]
@shmerl Thanks for taking a look. There is still some room for optimization, and some of that might still happen within that MR. But it's already a massive changeset with ~30 commits so right now the main focus is to get it integrated. I think it's a big step that ACO can now handle all stages, at least.

**Comment by SpookySkeletons** [comment]
The build with aco-tess mesa seems to be failing on my end due to warnings becoming errors. This happening on your end?

Bah, just int-conversion warnings from the mesa aco-tess branch somewhere in the gentoo mesa-9999 ebuild. Applying the branch diff to master works fine for me.

**Comment by shmerl** [comment]
@Venemo : what would be the next focus after it's merged, performance optimizations or plumbing aco for OpenGL?

**Comment by Venemo** [comment]
@shmerl Right now we're focusing on closing the feature gap between ACO and RADV's current LLVM backend. This includes 8/16-bit support (storage and I/O), tessellation and a few other things.

Plumbing ACO into RadeonSI (the OpenGL driver) is also something that I'd like to do, but it's not as trivial as it sounds. First, we need to take some steps to unify how RADV and RadeonSI do some things.

**Comment by Venemo** [comment]
The `aco-tess` branch is now merged. It should give you guys equal functionality to what you currently get from LLVM, and it should work accross all supported hardware, GFX6-10.

Unfortunately it is currently slightly less efficient than the LLVM implementation. We'll improve it by adding support for shader I/O in the NIR load/store vectorizer. We are also looking into making it use less LDS space (for the sake of sanity, I used the same LDS layout as LLVM, for the initial version).

**Comment by shmerl** [comment]
Do you plan to start working on OpenGL support in aco soon, or that's not yet scheduled? It would be good to avoid hangs caused by llvm with radeonsi.

**Comment by daniel-schuermann** [comment]
> Do you plan to start working on OpenGL support in aco soon, or that's not yet scheduled? It would be good to avoid hangs caused by llvm with radeonsi.

Yes, but it doesn't have the highest priority. For mesa-20.2, there isn't much more to be expected as it's around the corner (We'll try to get NGG GS in, though). For mesa-20.3, we've currently only planned to add float16 vectorization, but at least some radeonSI work (e.g. compute shaders or maybe more) is likely. Then, there is also the RT-extension which is good chunk of work and we should get started rather sooner than later to not lag too far behind :)

**Comment by darkbasic** [comment]
What would need to be done in order to get image support in clover for radeonsi? I've heard it needs aco

**Comment by Venemo** [comment]
@darkbasic It doesn't need ACO, and there is already an open MR for it: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/7231

**Comment by shmerl** [comment]
Are there any news / updates on plans or on-going work on ACO for radeonsi? Recent GPU hangs likely caused by radeonsi+llvm kind of reminded me about this.
